rts/motoko-rts/src/gc/generational.rs

```diff
@@ -0,0 +1,499 @@
+        let address = object as usize;
+        match self.strategy {
+            Strategy::Young => address >= self.heap.limits.last_free,
+            Strategy::Full => address >= self.heap.limits.base,
+        }
+    }
+
+    unsafe fn compact_phase(&mut self) {
+        if self.is_compaction_beneficial() {
+            self.thread_initial_phase();
+            self.move_phase();
+        }
+    }
+
+    fn is_compaction_beneficial(&self) -> bool {
+        self.survival_rate() < 0.95
+    }
+
+    fn generation_size(&self) -> usize {
+        let base = match self.strategy {
+            Strategy::Young => self.heap.limits.last_free,
+            Strategy::Full => self.heap.limits.base,
+        };
+        self.heap.limits.free - base
+    }
+
+    fn survival_rate(&self) -> f64 {
+        self.marked_space as f64 / self.generation_size() as f64
+    }
+
+    unsafe fn thread_initial_phase(&mut self) {
+        self.thread_all_backward_pointers();
+
+        // For static roots, also forward pointers are threaded.
+        // Therefore, this must happen after the heap traversal for backwards pointer threading.
+        self.thread_static_roots();
+
+        if (*self.heap.roots.continuation_table_ptr_loc).is_ptr() {
+            self.thread(self.heap.roots.continuation_table_ptr_loc);
+        }
+
+        // For the young generation GC run, the forward pointers from the old generation must be threaded too.
+        if self.strategy == Strategy::Young {
+            self.thread_old_generation_pointers();
+        }
+    }
+
+    unsafe fn thread_static_roots(&self) {
+        let root_array = self.heap.roots.static_roots.as_array();
+        for i in 0..root_array.len() {
+            let object = root_array.get(i).as_obj();
+            debug_assert_eq!(object.tag(), TAG_MUTBOX);
+            debug_assert!((object as usize) < self.heap.limits.base);
+            self.thread_root_mutbox_fields(object as *mut MutBox);
+        }
+    }
+
+    unsafe fn thread_root_mutbox_fields(&self, mutbox: *mut MutBox) {
+        let field_addr = &mut (*mutbox).field;
+        if pointer_to_dynamic_heap(field_addr, self.heap.limits.base) {
+            self.thread(field_addr);
+        }
+    }
+
+    unsafe fn thread_all_backward_pointers(&mut self) {
+        let mut bitmap_iter = iter_bits();
+        if self.strategy == Strategy::Young {
+            bitmap_iter.advance(self.heap.limits.last_free as u32);
+        }
+        let mut bit = bitmap_iter.next();
+        while bit != BITMAP_ITER_END {
+            let object = (bit * WORD_SIZE) as *mut Obj;
+            self.thread_backward_pointer_fields(object);
+            bit = bitmap_iter.next();
+        }
+    }
+
+    unsafe fn thread_backward_pointer_fields(&mut self, object: *mut Obj) {
+        debug_assert!(object.tag() < TAG_ARRAY_SLICE_MIN);
+        visit_pointer_fields(
+            &mut (),
+            object,
+            object.tag(),
+            self.heap.limits.base,
```

Comment:
Again, for a young colletion, can we restrict this to `heap.limits.last_free`?
Similar to previous comment.

Reply:
Improved in [fcbb1a4](https://github.com/dfinity/motoko/pull/3495/commits/fcbb1a434dac240daaac1e059804a3cfaa3f1459).